nixos/modules/services/networking/newt.nix

```
@@ -0,0 +1,161 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.services.newt;
+in
+{
+  options = {
+    services.newt = {
+      enable = lib.mkEnableOption "Newt, user space tunnel client for Pangolin";
+      # needs to be changed when newt-go changes to fosrl-newt
+      package = lib.mkPackageOption pkgs "newt-go" { };
+
+      id = lib.mkOption {
+        type = with lib.types; nullOr str;
+        default = null;
+        description = ''
+          The Newt Id that will be used to communicate to Pangolin. This is generated on site creation in the dashboard.
+        '';
+      };
+      endpoint = lib.mkOption {
+        type = with lib.types; nullOr str;
+        default = null;
+        description = ''
+          The endpoint where both Gerbil and Pangolin reside in order to connect to the websocket. The url of your Pangolin dashboard.
+        '';
+      };
+      logLevel = lib.mkOption {
+        type = lib.types.enum [
+          "DEBUG"
+          "INFO"
+          "WARN"
+          "ERROR"
+          "FATAL"
+        ];
+        default = "INFO";
+        description = "The log level to use.";
+      };
+      # provide path to file to keep secrets out of the nix store
+      environmentFile = lib.mkOption {
+        type = with lib.types; nullOr path;
+        default = null;
+        description = ''
+          Path to a file containing sensitive environment variables for Newt. See https://docs.fossorial.io/Newt/overview#cli-args
+          These will overwrite anything defined in the config.
+          The file should contain environment-variable assignments like:
+          NEWT_ID=2ix2t8xk22ubpfy
+          NEWT_SECRET=nnisrfsdfc7prqsp9ewo1dvtvci50j5uiqotez00dgap0ii2
+        '';
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    assertions = [
+      {
+        assertion = cfg.environmentFile != null;
+        message = "services.newt.environmentFile must be provided when Newt is enabled.";
+      }
+    ];
+
+    systemd.services.newt = {
+      description = "Newt, user space tunnel client for Pangolin";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+
+      environment = {
+        HOME = "/var/lib/private/newt";
+      };
+      # the flag values will all be overwritten if also defined in the env file
+      script = ''
+        exec ${lib.getExe pkgs.newt-go} \
+        ${lib.optionalString (!isNull cfg.id) "--id ${cfg.id} \\"}
+        ${lib.optionalString (!isNull cfg.endpoint) "--endpoint ${cfg.endpoint} \\"}
+        --log-level ${cfg.logLevel}
+      '';
```

```suggestion
      script = "
        exec ${lib.getExe pkgs.newt-go} \\\n
        ${lib.optionalString (!isNull cfg.id) "--id ${cfg.id} \\\n"}
        ${lib.optionalString (!isNull cfg.endpoint) "--endpoint ${cfg.endpoint} \\\n"}
        --log-level ${cfg.logLevel}
      ";
```
Looks like this breaks if we use multi-line strings.

---

good one, fixed.